### PR TITLE
Fix error in constructor

### PR DESCRIPTION
**Pull request by bkamins**
Fixes 2 errors:
* access to `#undef`
* wrong handling of non 1-based source arrays

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -601,3 +601,21 @@ end
     @test popfirst!(x) == "1"
     @test x == ["2"]
 end
+
+@testset "constructor corner cases" begin
+    x = Vector{Any}(undef, 3)
+    y = PooledArray(x)
+    @test y isa PooledArray{Any}
+    @test all(iszero, y.refs)
```
Comment (nalimilan) [inline test/runtests.jl:610]: Maybe also test directly the expected API?
```suggestion
    @test !any(i -> isassigned(y, i), eachindex(y))
    @test all(iszero, y.refs)
```

File: test/runtests.jl [hunk 2]
```diff
@@ -601,3 +601,21 @@ end
...
+    @test isempty(y.pool)
+    @test isempty(y.invpool)
+
+    x[2] = "a"
+    for v in (x, OffsetVector(x, -5))
+        y = PooledArray(v)
+        @test y isa PooledArray{Any}
+        @test y.refs == [0, 1, 0]
```
Comment (nalimilan) [inline test/runtests.jl:621]:
```suggestion
        @test !isassigned(x, 1)
        @test x[2] == "a"
        @test !isassigned(x, 3)
        @test y.refs == [0, 1, 0]
```

**Review by nalimilan — APPROVED** [review]

File: test/runtests.jl (hunk 1 above)
Comment (bkamins) [reply test/runtests.jl:610]: OK - I have fixed your proposal to make it work.

**Comment by bkamins** [comment]
Thank you!
